### PR TITLE
ddl, meta: replace etcd mutex with kv distributed lock

### LINKED ISSUE #48546: add index failed with error “Error 1105 (HY000): mutex: session is expired”
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、run workload
2、add index for one table
3、inject pdleader io delay 1s 

case config：
      tag: "ha-test-add-index"
      workloads:
        - type: sysbench
          storage: s3://benchmark/sysbench_64_7000w
          db: sysbench_64_7000w
          tableNum: 64
          tableSize: 70000000
          threads: 64
          subType: oltp_read_write
          ignoreErrors: "all"
      cases:
        - name: ha_pd(leader)_io_delay(1s)
          faultType: iochaos_io_delay(volumePath:/var/lib/pd,path:/var/lib/pd/data/**/*,delay:1s,percent:100)
          selector: pd(leader)
          warmUpTime: 1m
          period: "@every 5m"
          faultDuration: 2m
          faultTotalRunTime: 30m
          ddl:
            addIndex:
              - db: "sysbench_64_7000w"
                table: "sbtest1"
                timeout: 60m
                times: 5
                index_col_name: "c"
                job_type: "ingest"
                admin_check: false
                count_check: false

### 2. What did you expect to see? (Required)
add index can success

### 3. What did you see instead (Required)
add index failed with error “Error 1105 (HY000): mutex: session is expired”

add index failed at 2023-11-12 14:14:05 (Error 1105 (HY000): mutex: session is expired)
operatorLogs:
[2023-11-12 14:10:07] ###### start adding index
alter table sbtest1 add index index_test_1699769407963 (c)
[2023-11-12 14:10:07] ###### wait for ddl job finish

### 4. What is your TiDB version? (Required)
git hash：700beafa79844b7b48dcba1c452ea3ff49d8f271

### LINKED ISSUE #48904: add index is always queueing due to task is blocked after tikv rolling restart
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
tidb_enable_dist_task='on'
1、run workload
2、add index
3、tikv rolling restart

### 2. What did you expect to see? (Required)
add index can success

### 3. What did you see instead (Required)
add index is always queueing due to task is blocked after do some fault test

### 4. What is your TiDB version? (Required)
git hash：8eb191303aca84982479c98407fd4e6bf158dc14

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #48546 close #48904

Problem Summary:

Previously, the distributed lock is implemented with etcd's `session.Mutex`. We have encountered several problems:

1. "mutex: session is expired" error mentioned in #48546. We need a way to enlarge the lease.
2. The mutex may be blocked forever when etcd members change during `lock()` and `unlock()`.
3. TBD

### What changed and how does it work?

Implement a general distributed lock with transaction.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
  image
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/meta/distributed_lock.go [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+				newLockCreated = true
+				return nil, nil
+			}
+			// Lock is valid, wait for a while.
+			return lockHeldErr, nil
+		})
+	if err != nil {
+		return err
+	}
+	if !intest.InTest && newLockCreated {
```
Comment (wjhuang2016) [inline pkg/meta/distributed_lock.go:189]: newLockCreated is always true here?

Comment (tangenta) [reply pkg/meta/distributed_lock.go:189]: No, `getLockInTxnWithRetry()` may encounter error without running `newLockCreated = true`. For example, the outer context is cancelled, or there is a decoding error on `lockContent`.

File: pkg/meta/distributed_lock.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+func (c lockContent) encode() ([]byte, error) {
+	return json.Marshal(c)
+}
+
+func (c *lockContent) decode(data []byte) error {
+	return json.Unmarshal(data, c)
+}
+
+// DistributedLock is a distributed lock implementation based on meta key transaction.
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:65]: Can you explain how do we implement lease timeout behaviour? For example, every clients check xx TS to know it's timed out and try to acquire the lock, and contention will be found when KV transaction commits?

**Comment by lance6716** [comment]
/review please check the files under /pkg/meta and give PR review comments like you are a distributed lock expert and golang expert

**Comment by ywqzzy** [comment]
/review please check the files under /pkg/meta and give PR review comments like you are a distributed lock expert and golang expert, please give more comments on whether the distributed lock will have bug on chaos environment. And tell us the limit of the distributed lock.

File: pkg/meta/distributed_lock.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+func (d *DistributedLock) renewLockLoop(ctx context.Context) {
+	ticker := time.NewTicker(d.lease / 3)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := d.renew(ctx)
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:130]: In order to survive the temporary network error, I think we should only exit when the error is "lock is unexpectedly removed" case.

File: pkg/meta/distributed_lock.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+			if err != nil {
+				logutil.BgLogger().Warn("renew lock failed", zap.Error(err))
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:132]: b.ctx has attached the category, I guess you forget to use the logger of it?

File: pkg/meta/distributed_lock.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-d.outerCtx.Done():
+			// We only check the outer context error after retry 10 times,
+			// so that the lock can be released even if the outer context is canceled.
+			if retryCnt >= 10 {
+				return d.outerCtx.Err()
+			}
+		default:
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:246]: time.Sleep is not good, should use case <- time.After()

File: pkg/meta/distributed_lock.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+	err = fn(txn)
+	if err != nil {
+		txn.Rollback()
+		return err
+	}
+	return txn.Commit()
+}
+
+func decodeLockContent(data []byte) (*lockContent, error) {
+	if len(data) == 0 {
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:305]: in which case len(data) is 0 and txn.Get has no error?

File: pkg/meta/distributed_lock.go (hunk 3 above)
Comment (tangenta) [reply pkg/meta/distributed_lock.go:130]: `d.renew` itself also has retry logic.

File: pkg/meta/distributed_lock.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+	}()
+	return d.getLockInTxnWithRetry(context.Background(),
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			if lc == nil || lc.InstanceID != d.instanceID {
+				logutil.Logger(d.outerCtx).Warn("lock is unexpectedly removed", zap.String("instanceID", d.instanceID))
+				return nil, nil
+			}
+
+			// Lock is exist, remove it.
+			err := txn.Del(d.lockName)
```
[the remark this answers is not in the document]
Comment (tangenta) [reply pkg/meta/distributed_lock.go:226]: Yes.

File: pkg/meta/distributed_lock.go (hunk 6 above)
Comment (tangenta) [reply pkg/meta/distributed_lock.go:305]: That is why I change the implementation of `DataTxn.Get()`:

```
if kv.ErrNotExist.Equal(err) {
	err = nil
}
```

File: pkg/meta/distributed_lock.go [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-d.outerCtx.Done():
+			// We only check the outer context error after retry 10 times,
+			// so that the lock can be released even if the outer context is canceled.
+			if retryCnt >= 10 {
+				return d.outerCtx.Err()
+			}
+		case <-backoffTimer.C:
+			var noRetry bool
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:246]: these logic should be moved out of `select`, and let `case <-backoffTimer.C` be an empty case. If d.outerCtx is done we have little chance to enter `case <-backoffTimer.C`

File: pkg/meta/distributed_lock.go [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+			return errors.Trace(err), nil
+		})
+}
+
+func (d *DistributedLock) getLockInTxnWithRetry(ctx context.Context, fn lockHandleFunc) error {
+	retryCnt := 0
+	backoffTimer := time.NewTimer(d.backoff)
+	defer backoffTimer.Stop()
+	for {
+		backoffTimer.Reset(d.backoff)
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:235]: See the function comment of Reset, it's hard to use it correctly. maybe we can directly use time.After 😂

File: pkg/meta/distributed_lock.go [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-d.outerCtx.Done():
+			// We only check the outer context error after retry 10 times,
+			// so that the lock can be released even if the outer context is canceled.
+			if retryCnt >= 10 {
+				return d.outerCtx.Err()
+			}
```
Comment (lance6716) [inline pkg/meta/distributed_lock.go:244]: and here we should still check backoffTimer and ctx

File: pkg/ddl/ingest/backend.go [hunk 11]
```diff
@@ -193,28 +188,28 @@ func (bc *litBackendCtx) Flush(indexID int64, mode FlushMode) (flushed, imported
...
 				logutil.Logger(bc.ctx).Info("release distributed flush lock success", zap.Int64("jobID", bc.jobID))
 			}
-			err = se.Close()
-			if err != nil {
-				logutil.Logger(bc.ctx).Warn("close session error", zap.Error(err))
-			}
 		}()
 	}
+
 	err = bc.unsafeImportAndReset(ei)
```
Comment (lance6716) [inline pkg/ddl/ingest/backend.go:213]: I don't understand how distributed lock solves the problem. If node A acquired the lock, started an import, then lock is expired and node B now acquired the lock. A has no way to cancel its running import task after B starts? If we tolerate two nodes run import task simultaneously for a short time, we can use more simple way to implement it?

Comment (tangenta) [reply pkg/ddl/ingest/backend.go:213]: Could you explain the simple way in details?

> cancel the task when the task owner node is not itself

Then we need a way to persist the state of "task owner node". This PR stores this state to a meta key.

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/meta/distributed_lock.go [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+				lc := newLockContent(startTS, d.instanceID)
+				data, err := lc.encode()
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				err = txn.Set(d.lockName, data)
+				if err != nil {
+					return errors.Trace(err), nil
+				}
+				return nil, nil
```
Comment (Benjamin2037) [inline pkg/meta/distributed_lock.go:212]: Here, the lock should be reset successfully.  
```suggestion
				return ture nil
```

File: pkg/meta/distributed_lock.go [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+			lockIsHeld = true
+			return lockHeldNonBlockingErr, nil
+		})
+	if err != nil {
+		return false, err
+	}
+	if lockIsHeld {
+		return false, nil
+	}
+	d.runRenewLoop()
```
Comment (Benjamin2037) [inline pkg/meta/distributed_lock.go:224]: if require lock is successful, we should not go into runRenewLoop?

File: pkg/meta/distributed_lock.go [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+
+var (
+	// lockHeldErr is returned when the lock is held by another routine.
+	lockHeldErr = errors.Errorf("lock is held")
+	// lockHeldNonBlockingErr is used to indicate that the lock is held and the caller should not retry.
+	lockHeldNonBlockingErr = errors.Errorf("lock is held, non-blocking")
+)
+
+// Lock locks the distributed lock.
+func (d *DistributedLock) Lock() error {
```
Comment (Benjamin2037) [inline pkg/meta/distributed_lock.go:168]: For this implement, seems try lock and lock are almost the same. lock should stop and wait until unlock to release blocked applier.

File: pkg/meta/distributed_lock.go (hunk 13 above)
Comment (tangenta) [reply pkg/meta/distributed_lock.go:224]: We need to update the lease while acquiring the lock.

File: pkg/meta/distributed_lock.go (hunk 14 above)
Comment (tangenta) [reply pkg/meta/distributed_lock.go:168]: Yes. This is the difference between `Lock` and `TryLock`.

**Review by Benjamin2037 — APPROVED** [review]
rest LGTM

**Pull request closed without merging**